File: src/main/java/sage/transfer/MidForwards.java
package sage.transfer;

import java.util.LinkedList;
import java.util.List;

import sage.entity.Tweet;
import sage.web.context.Json;

public class MidForwards {
  private List<Long> ids = new LinkedList<>();
  private List<String> msgs = new LinkedList<>();

  MidForwards() {}

  public MidForwards(Tweet directForward) {
    addForward(directForward);
    MidForwards formerMidForwards = directForward.midForwards();
    if (formerMidForwards != null) {
      ids.addAll(formerMidForwards.getIds());
      msgs.addAll(formerMidForwards.getMsgs());
    }
  }

  MidForwards addForward(Tweet forward) {
    Tweet t = forward;
    ids.add(t.getId());
    String asString = "@" + t.getAuthor().getName() + "#" + t.getAuthor().getId() + " : " + t.getContent();
    msgs.add(asString);
    return this;
  }

  public MidForwards removeById(Long idToRemove) {
    int idx = getIds().indexOf(idToRemove);
    getIds().remove(idx);
    getMsgs().remove(idx);
    return this;
  }

  public static MidForwards from(Tweet tweet) {
    return Json.object(tweet.getMidForwardsJson(), MidForwards.class);
  }

  public static MidForwards fromJson(String json) {
    return Json.object(json, MidForwards.class);
  }

  public String toJson() {
    return Json.json(this);
  }

  public List<Long> getIds() {
    return ids;
  }
  public List<String> getMsgs() {
    return msgs;
  }
}

File: src/main/kotlin/sage/transfer/MidForwards.kt
package sage.transfer

import sage.annotation.KotlinNoArg
import java.util.LinkedList

import sage.entity.Tweet
import sage.web.context.Json

@KotlinNoArg
class MidForwards {
  var xs: MutableList<MidForward> = LinkedList()

  constructor()

  constructor(directForward: Tweet) {
    addForward(directForward)
    val formerMidForwards = directForward.midForwards()
    if (formerMidForwards != null) { xs.addAll(formerMidForwards.xs) }
  }

  private fun addForward(forward: Tweet): MidForwards {
    val t = forward
    xs.add(MidForward(t.id, t.author.id, t.author.name, t.content))
    return this
  }

  fun removeById(idToRemove: Long): MidForwards {
    val idx = xs.indexOfFirst { it.id == idToRemove }
    if (idx >= 0) { xs.removeAt(idx) }
    return this
  }

  fun toJson(): String {
    return Json.json(this)
  }

  companion object {

    fun from(tweet: Tweet): MidForwards {
      return Json.`object`(tweet.midForwardsJson!!, MidForwards::class.java)
    }

    fun fromJson(json: String): MidForwards {
      return Json.`object`(json, MidForwards::class.java)
    }
  }
}
